---
layout: home
permalink: /
image:
  feature: Phyloref-owl-large.jpeg
---


## The problem

Scientists across the globe collect ever more data about ever more
characteristics of life's diversity on earth. This breadth of data
becomes truly powerful only when data from the many different sources,
studies, and researchers can be integrated, queried, and aggregated
by machines, with precision, and at scale.

However, to this day the by far most common way to link organismal
data relies on organism names based in Linnaean nomenclature. As a
consequence, linking organismal data is fraught with the following two
major issues:

1. Linnaean names are not simply labels, but carry meaning
   ("semantics") by defining which organisms are and which are not
   included. These semantics are inaccessible to machines. Worse yet,
   they can change over time, and frequently differ between those who
   coin a name and those who apply it.
2. Because of the nomenclatural process, there are many groups of
   organisms that do not yet and may never have a Linnaean name. In a
   name-based system, data without an attached name are dark. Unlike
   the genes encoded in a genome (whose naming is also governed by a
   nomenclatural process), there is no reason to predefine the groups
   for which biological knowledge is worth collecting or
   communicating.

## About this project

We aim to create a standard mechanism for defining biological clades
with precise and fully machine-processible semantics. We call this
mechanism _phyloreferencing_, in analogy to the ubiquitous mechanism
of _georeferencing_ for geographic places. Instead of geocoordinates
as the computable reference expression, _phyloreferences_ use a
machine-interpretable definition of the unique pattern of shared
evolutionary descent that distinguishes a group from all others. Such
a group can include many species, or only a population within a
species, or, for example, those carrying the descendants of a shared
ancestral gene.

In contrast to authoritative nomenclatural naming, our goal for
phyloreferences is that users can construct them instantly for any
group of shared evolutionary descent for which they wish to
communicate data or discoveries.

Our work builds on more than a decadeâ€™s worth of theoretical and
applied research into phylogenetic taxonomy. In addition to developing
the phyloreferencing standard, as part of this project we will also
create classroom-ready teaching materials and curricula for training
biology students in the theoretical underpinnings and practical
applications of phylogenetic nomenclature.

### Funding

<div style="float: right; max-width: 128px; margin-top:
-10px;"><img src="http://www.nsf.gov/images/logos/nsf1.jpg"/></div>
This project is funded by the US National Science Foundation through
collaborative grants
[DBI-1458484](http://www.nsf.gov/awardsearch/showAward?AWD_ID=1458484)
and
[DBI-1458604](http://www.nsf.gov/awardsearch/showAward?AWD_ID=1458604)
to Hilmar Lapp (Duke University) and Nico Cellinese (University of
Florida), respectively. The proposal text is available online: <cite>Cellinese, Nico; Lapp, Hilmar (2015): An Ontology-Based System for Querying Life in a Post-Taxonomic Age. figshare. <https://dx.doi.org/10.6084/m9.figshare.1401984></cite>

## People

### Project personnel

* [Nico Cellinese](http://www.flmnh.ufl.edu/museum-voices/nico-cellinese/) (PI, University of Florida)
* [Hilmar Lapp](http://orcid.org/0000-0001-9107-0714) (PI, Duke University)
* [Gaurav Vaidya](http://www.ggvaidya.com) (University of Florida)

### Advisory Board

* Jim Balhoff (RTI)
* David Baum (University of Wisconsin, Madison)
* Holly Bik (University of California, Riverside)
* Christopher Mungall (Lawrence Berkeley National Laboratory)
* Susan Perkins (American Museum of Natural History)
* Michael Sanderson (University of Arizona)
